cmd/soroban-rpc/internal/db/transaction.go

```diff
@@ -0,0 +1,336 @@
+	log        *log.Entry
+	db         db.SessionInterface
+	stmtCache  *sq.StmtCache
+	passphrase string
+
+	ingestMetric, insertMetric, countMetric prometheus.Observer
+}
+
+func NewTransactionReader(log *log.Entry, db db.SessionInterface, passphrase string) TransactionReader {
+	return &transactionHandler{log: log, db: db, passphrase: passphrase}
+}
+
+func (txn *transactionHandler) InsertTransactions(lcm xdr.LedgerCloseMeta) error {
+	start, mid := time.Now(), time.Now()
+	txCount := lcm.CountTransactions()
+	L := txn.log.
+		WithField("ledger_seq", lcm.LedgerSequence()).
+		WithField("tx_count", txCount)
+
+	defer func() {
+		if txn.ingestMetric != nil {
+			txn.ingestMetric.Observe(time.Since(start).Seconds())
+			txn.insertMetric.Observe(time.Since(mid).Seconds())
+			txn.countMetric.Observe(float64(txCount))
+		}
+	}()
+
+	if txn.stmtCache == nil {
+		return errors.New("TransactionWriter incorrectly initialized without stmtCache")
+	} else if txCount == 0 {
+		return nil
+	}
+
+	reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(txn.passphrase, lcm)
+	if err != nil {
+		return errors.Wrapf(err,
+			"failed to open transaction reader for ledger %d",
+			lcm.LedgerSequence())
+	}
+
+	transactions := make(map[xdr.Hash]ingest.LedgerTransaction, txCount)
+	for i := 0; i < txCount; i++ {
+		tx, err := reader.Read()
+		if err != nil {
+			return errors.Wrapf(err, "failed reading tx %d", i)
+		}
+
+		// For fee-bump transactions, we store lookup entries for both the outer
+		// and inner hashes.
+		if tx.Envelope.IsFeeBump() {
+			transactions[tx.Result.InnerHash()] = tx
+		}
+		transactions[tx.Result.TransactionHash] = tx
+	}
+
+	mid = time.Now()
+	query := sq.Insert(transactionTableName).
+		Columns("hash", "ledger_sequence", "application_order")
+	for hash, tx := range transactions {
+		query = query.Values(hash[:], lcm.LedgerSequence(), tx.Index)
+	}
+	_, err = query.RunWith(txn.stmtCache).Exec()
+
+	L.WithError(err).
+		WithField("duration", time.Since(start)).
+		WithField("sql_duration", time.Since(mid)).
+		Infof("Ingested %d transaction lookups", len(transactions))
+
+	return err
+}
+
+func (txn *transactionHandler) RegisterMetrics(ingest, insert, count prometheus.Observer) {
+	txn.ingestMetric = ingest
+	txn.insertMetric = insert
+	txn.countMetric = count
+}
+
+// trimTransactions removes all transactions which fall outside the ledger retention window.
+func (txn *transactionHandler) trimTransactions(latestLedgerSeq uint32, retentionWindow uint32) error {
+	if latestLedgerSeq+1 <= retentionWindow {
+		return nil
+	}
+
+	cutoff := latestLedgerSeq + 1 - retentionWindow
+	_, err := sq.StatementBuilder.
+		RunWith(txn.stmtCache).
+		Delete(transactionTableName).
+		Where(sq.Lt{"ledger_sequence": cutoff}).
+		Exec()
+	return err
+}
+
+// GetLedgerRange pulls the min/max ledger sequence numbers from the database.
+func (txn *transactionHandler) GetLedgerRange(ctx context.Context) (ledgerbucketwindow.LedgerRange, error) {
```

Comment:
I just realized that we could maintain the ledger range as in-memory fields that are updated by the ingestion thread:

Upon start up you query the oldest and newest ledger. Then whenever we successfully ingest a new ledger we increment the newest ledger field. And, whenever we trim an old ledger we increment the oldest ledger field.

That would allow us to reduce the amount of db queries we make.

Reply:
Nice!! I had a similar idea but not to this level of depth because it was before I knew about trimming. I propose that we do this in a later PR (note that this PR doesn't target `main`) to isolate the changes as well as enhance testing!

Reply:
sure, sounds good!